servers/netserver/src/ip/icmp.cpp

```
@@ -0,0 +1,290 @@
+				});
+		}
+
+		if(self->timestamp_) {
+			auto truncated = ctrl.message(SOL_SOCKET, SCM_TIMESTAMP, sizeof(struct timeval));
+			if(!truncated)
+				ctrl.write(element->recvTimestamp);
+		}
+
+		if(self->ipRecvTtl_) {
+			auto truncated = ctrl.message(SOL_IP, IP_TTL, sizeof(int));
+			if(!truncated)
+				ctrl.write<int>(element->packet->header.ttl);
+		}
+
+		if(self->ipRetOpts_) {
+			arch::dma_buffer_view options = element->packet->header_view().subview(sizeof(Ip4Packet::Header));
+
+			if(options.size()) {
+				auto truncated = ctrl.message(SOL_IP, IP_RETOPTS, options.size());
+				if(!truncated)
+					ctrl.write_buffer(options);
+			}
+		}
+
+		co_return RecvData{ctrl.buffer(), copy_size, sizeof(addr), 0};
+	}
+
+	static async::result<frg::expected<protocols::fs::Error, size_t>> sendmsg(void *,
+			const char *creds, uint32_t flags,
+			void *data, size_t len,
+			void *addr_ptr, size_t addr_size,
+			std::vector<uint32_t> fds, struct ucred) {
+		using arch::endian;
+		(void) creds;
+		(void) flags;
+		(void) fds;
+
+		assert(!flags);
+
+		if(addr_size < sizeof(sockaddr_in))
+			co_return Error::illegalArguments;
+
+		sockaddr_in target;
+		memcpy(&target, addr_ptr, std::min(addr_size, sizeof(sockaddr_in)));
+
+		auto ti = co_await ip4().targetByRemote(arch::convert_endian<endian::big, endian::native>(target.sin_addr.s_addr));
+		if (!ti)
+			co_return protocols::fs::Error::netUnreachable;
+
+		auto error = co_await ip4().sendFrame(std::move(*ti),
+			data, len,
+			static_cast<uint16_t>(IpProto::icmp));
+
+		if (error != protocols::fs::Error::none)
+			co_return error;
+
+		co_return len;
+	}
+
+	static async::result<frg::expected<protocols::fs::Error, protocols::fs::PollWaitResult>>
+	pollWait(void *obj, uint64_t past_seq, int mask, async::cancellation_token cancellation) {
+		auto self = static_cast<IcmpSocket *>(obj);
+		(void)mask; // TODO: utilize mask.
+
+		assert(past_seq <= self->_currentSeq);
```

This should return an error.